src/nvim/tui/tui.c

```diff
@@ -875,6 +875,53 @@ static void cursor_goto(UI *ui, int row, int col)
   ugrid_goto(grid, row, col);
 }
 
+static void print_spaces(UI *ui, int width)
+{
+  TUIData *data = ui->data;
+  UGrid *grid = &data->grid;
+
+  out(ui, data->space_buf, (size_t)width);
+  grid->col += width;
+  if (data->immediate_wrap_after_last_column) {
+    // Printing at the right margin immediately advances the cursor.
+    final_column_wrap(ui);
+  }
+}
+
+/// Move cursor to the position given by `row` and `col` and print the character in `cell`.
+/// This allows the grid to assume wider ambiguous-width characters than the host terminal.
+///
+/// @param is_doublewidth  whether the character is double-width on the grid.
+///                        If true and the character is ambiguous-width, clear two cells.
+static void print_cell_at_pos(UI *ui, int row, int col, UCell *cell, bool is_doublewidth)
+{
+  TUIData *data = ui->data;
+  UGrid *grid = &data->grid;
+
+  if (grid->row == -1 && cell->data[0] == NUL) {
+    // If cursor needs to repositioned and there is nothing to print, don't move cursor.
+    return;
+  }
+
+  cursor_goto(ui, row, col);
+
+  bool is_ambiwidth = utf_ambiguous_width(utf_ptr2char(cell->data));
+  if (is_ambiwidth && is_doublewidth) {
+    // Clear the two screen cells.
+    // If the character is single-width in the host terminal it won't change the second cell.
+    update_attrs(ui, cell->attr);
+    print_spaces(ui, 2);
+    cursor_goto(ui, row, col);
+  }
+
+  print_cell(ui, cell);
+
+  if (is_ambiwidth) {
```

Comment:
Not sure if I should add a ` && is_doublewidth` check here.
If I do not add a ` && is_doublewidth` check here, this matches Vim's behavior better, and also makes cursor position correct on a line with emoji modifiers (ref #7151), but performance is possibly not ideal.
If I add a ` && is_doublewidth` check here, cursor position is still broken on a line with emoji modifiers, but performance can be better.